### ISSUE TITLE
Live Validation?

### ISSUE THREAD

**Issue by MentalGear** [ciscoheat/sveltekit-superforms#101]
Thx for this library. I was wondering if there's also a way to have live validation, e.g. onBlur of a field, instead of submit of the whole form ?

**Comment by ciscoheat** [comment]
Hey, what a coincidence, just a few hours ago I put up a roadmap page describing that this will be the next big feature: https://superforms.vercel.app/roadmap

**Comment by MentalGear** [comment]
"Like minds" :) ! I must say specifically the universal validation and the excellent documentation is what makes this a highly valuable lib !

One con: zod's chaining syntax is not really readable. Would it be hard to plugin another lib like vest or superstruct ?

**Comment by MentalGear** [comment]
Thank you for explaining zod's introspective features and why it's the best lib to use in this case !

I have used [use-vest](https://github.com/enyo/use-vest) previously which is also terrific, but lacks docs->users->LTS, yet has live validation, which could maybe inspire you.

Funny again that you mention `async` ( we must really be synced :P ). In my previous reply I was going to ask for it, but thought there's no way to add it yet with superforms - or is there before 0.8 ?

**Comment by ciscoheat** [comment]
Synchronicity is a good sign. :) I'm working on live validation now, making pretty good progress so I think it won't be more than a few weeks away, tops.

Async validation is already possible, not just in real time, but both the Zod schema and the Superforms [validators](https://superforms.vercel.app/concepts/client-validation#validators) can be used asynchronously.

**Comment by MentalGear** [comment]
Awesome!

May I assume "username async validation" could work like this ?
- client: username put in form, submits
- server: check username, returns either success/error as form response

**Comment by ciscoheat** [comment]
That's what happens without async, to make the client check the username it can be done like this:

```ts
const { form, ... } = superForm(data.form, {
  validators: {
    async username(username) {
      const request = await fetch('/check-username', ...)
      const data = await request.json()
      return data === false ? 'Username already exists' : null
    }
  }
})

**Comment by MentalGear** [comment]
Neat, maybe also a good example to have in the docs since it's a quite common use-case. :)

Btw: I just noticed that Superforms does no client-side zod validation, only html5 validation. I assume this will also be coming in 0.8 then, right ? :)

**Comment by ciscoheat** [comment]
Zod client-side validation is already supported, you need to specify it in the `validators` option. Described here: https://superforms.vercel.app/concepts/client-validation

**Comment by MentalGear** [comment]
Having to define a schema on the serverside and then again on the clientside runs counter to the whole SSOT "single source of truth" concept.

<img width="1083" alt="Screenshot 2023-04-19 at 12 26 29" src="https://user-images.githubusercontent.com/2837147/233047481-0524c131-77a7-404f-b72f-5fb83d632081.png">

It be great if the lib could just reuse the schema in the validator automatically to have a single setup., and also automatically combine html5 validators (no JS) and zod validators.

**Comment by ciscoheat** [comment]
It does this already. Import the schema to the client, then do:

```ts
const { form, errors, enhance } = superForm(data.form, {
  validators: schema
});
```

Automatic reuse cannot be made because it would mean importing Zod to the client even when no client-side validation exist, which bloats the client bundle too much. The Superforms validators is a lightweight alternative to doing that.

**Comment by hmiiro** [comment]
Its actual mentioned in the docs, i thought it wasn't there but on second check its there immediately under the validators section
```
validators: AnyZodObject | {
  field: (value) => string | string[] | null | undefined;
}
```

**Comment by ciscoheat** [comment]
Removing the enhancement tag since the feature is tracked in #66

**Comment by ciscoheat** [comment]
With the newly released v0.8, live validation is now possible! Announcement: https://github.com/ciscoheat/sveltekit-superforms/discussions/106

**Comment by MentalGear** [comment]
Awesome, that was fast! Is there an example available or should it be "just working" without further config?
